python/ray/data/dataset.py

```diff
@@ -2407,6 +2407,10 @@ def iter_batches(
         batch_size: Optional[int] = None,
         batch_format: str = "native",
         drop_last: bool = False,
+        shuffle: bool = False,
+        shuffle_buffer_capacity: Optional[int] = None,
+        shuffle_buffer_min_size: Optional[int] = None,
```

Comment:
I originally wanted to suggest `shuffle_buffer_blocks`, but with automatic block splitting this may not be something we'd expect users to calculate. Similar thoughts apply to the existing `prefetch_blocks`.

What if we just exposed a single `shuffle_buffer_size` that combines the functionality of both `shuffle_buffer_min_size` and `shuffle_buffer_capacity`?

Reply:
My feeling is also that `shuffle_buffer_min_size` will be common to set and `shuffle_buffer_capacity` is more of an advanced setting, but would love to get your thoughts on this @clarkzinzow

Reply:
**TL;DR:** How about we have a single `shuffle_buffer_min_size` argument for now, and introduce the planned `buffer_capacity` argument (which would replace `prefetch_blocks` and apply to both shuffling and non-shuffling batchers) in one of the follow-up PRs?

## Context

It's advantageous to have these as separate arguments because they control different things.
1. `shuffle_buffer_min_size` controls the minimum size of the buffer before random batches can be extracted; **this controls the randomness of the shuffle**.
2. `shuffle_buffer_capacity` controls the maximum size that the buffer can achieve before we start applying backpressure to the producer (e.g. stalling prefetching); **this controls the memory size limit for the shuffle buffer**.

Increasing (1) improves shuffle quality at the cost of longer latency-to-batch, while increasing (2) improves throughput at the cost of higher memory utilization; e.g. you could imagine keeping `shuffle_buffer_min_size` fixed at 50k samples since that randomness suffices, while increasing `shuffle_buffer_capacity` to 250k samples in order to get higher throughput via more aggressive prefetching, up to the CPU memory limit of the trainer. This is why these are exposed as separate knobs in TensorFlow Datasets via a prefetching buffer size and shuffling buffer size (see [SO answer](https://stackoverflow.com/a/47025850)) and in the [Petastorm Torch dataloader](https://petastorm.readthedocs.io/en/latest/api.html#petastorm.spark.spark_dataset_converter.SparkDatasetConverter.make_torch_dataloader) (although Petastorm only exposes a `shuffle_queue_capacity` and sets `shuffle_buffer_min_size = shuffle_queue_capacity - 1` and `shuffle_buffer_capacity = shuffle_buffer_min_size + batch_size`).

However, for this first PR, `shuffle_buffer_capacity` doesn't do much because prefetching is still controlled by `prefetch_blocks` and prefetching + shuffle-batching is synchronous. The eventual plan is to consolidate `shuffle_buffer_capacity` and `prefetch_blocks` into a single `buffer_capacity` argument (used for both the shuffling batcher and the regular batcher), where we aggressively prefetch blocks until the `buffer_capacity` is hit, at which point backpressure is applied; the plan is to do this in a follow up PR to this first shuffle PR, once we start moving shuffled and non-shuffled batch into the background, where prefetching already operates.

How about we have a single `shuffle_buffer_min_size` argument for now, and introduce the planned `buffer_capacity` argument (which would replace `prefetch_blocks` and apply to both shuffling and non-shuffling batchers) in one of the follow-up PRs?

Reply:
@matthewdeng @krfricke I've made this update, removing the `shuffle_buffer_capacity` argument, lmk what y'all think!

Reply:
Starting with just `shuffle_buffer_min_size` in this PR sounds good to me!

For discussion purposes, my understanding is a bit different. 

> you could imagine keeping shuffle_buffer_min_size fixed at 50k samples since that randomness suffices, while increasing shuffle_buffer_capacity to 250k samples in order to get higher throughput via more aggressive prefetching,

In this example, wouldn't the randomness increase because the buffer fills to 250k, dequeues to 50k, and repeats? 

Whereas in TensorFlow `dataset.shuffle(50_000).prefetch(250_000)` would have a shuffle buffer that's always at 50k, and 250k values will be dequeued (and enqueued) as part of the prefetch.

Reply:
For shuffle_buffer_min_size: if it's controlling randomness, can the "buffer" in the name be dropped? It's pretty much read like a request of memory allocation.

For unifying prefetch_blocks into buffer_capacity: I'm not sure it always makes sense. Aggressively prefetching blocks may not help if the consumption is the bottleneck.